[FLINK-1166] Add qa-check.sh tool

rmetzger opened the pull request:
The ./tools/qa-check.sh tool allows developers to check for common errors before opening a pull request.
The qa-checks are designed to be easily extensible by further checks.


I'm currently working on some tooling to automatically execute the tool on new pull requests (or on request).

uce commented:
Very nice. Will try it out now :)

mxm commented:
The script builds Flink twice and therefore takes a while. I don't know if we can force users to execute it before making pull requests. However, it would be good to get an automatic report for open pull requests using this script.

uce commented:
The idea is to run this automatically or on request, which should be fine.

mxm commented on tools/qa-check.sh:
```diff
@@ -0,0 +1,175 @@
+if [ ! -d  "flink" ] ; then
+	echo "There is no flink copy in the workdir. Cloning flink"
+	git clone https://github.com/apache/flink.git flink
+	cd flink
```
This `cd flink` is not necessary.

rmetzger commented:
Its not completely building flink, its only generating javadocs or compiling for getting the compiler warnings.

The main purpose of the script is to be executed automatically.

uce commented on tools/qa-check.sh:
```diff
@@ -0,0 +1,175 @@
+}
+
+############ Files in lib ############
+BUILD_MVN_COMMAND="mvn clean package -DskipTests"
```
we could also skip Javadocs here with `-Dmaven.javadoc.skip=true`

mxm commented:
I just ran the script on your pull request ;)

> Flink QA-Check results:
> :+1: The number of javadoc errors was      402 and is now      402
> :-1: The change increases the number of compiler warnings from      104 to      314
> :+1: The number of files in the lib/ folder was      142 before the change and is now      142

> Test finished.
> Overall result: :-1:. Some tests failed. Please check messages above

I guess if the user does not rebase to the latest master, then negative QA test results are likely to occur. Also, warnings could be hidden when the number of warnings have increased on the master. Wouldn't it be better to compare against the base of the user's changes?

rmetzger commented:
Thank you for trying it out.
Ideally, with the bot in place the number of warnings will go down over time.

I'll address the comments in the source.

I'm not sure if the number of compiler warnings is correct here.

A third option would be to
a) check out the current master
b) get the reference counts on the master
c) apply the pull request as a patch to master (checking if patching is possible (basically testing if rebase is possible))
d) if rebase was possible, get the new counts.

rmetzger commented:
I found out why the qa-check has more compile errors compared to the master: I've instructed the compiler to report them all ;)

I've addressed all comments in the pull request. I'm going to merge it soon if there are no other comments.
